validate.go

```diff
@@ -182,39 +182,39 @@ func checkProcess(process rspec.Process, rootfs string) {
 }
 
 //Linux only
-func checkLinux(spec rspec.Linux, hostname string, rootfs string) {
+func checkLinux(spec rspec.Spec) {
 	utsExists := false
 
-	if len(spec.UIDMappings) > 5 {
+	if len(spec.Linux.UIDMappings) > 5 {
 		logrus.Fatalf("Only 5 UID mappings are allowed (linux kernel restriction).")
 	}
-	if len(spec.GIDMappings) > 5 {
+	if len(spec.Linux.GIDMappings) > 5 {
 		logrus.Fatalf("Only 5 GID mappings are allowed (linux kernel restriction).")
 	}
 
-	for index := 0; index < len(spec.Namespaces); index++ {
-		if !namespaceValid(spec.Namespaces[index]) {
-			logrus.Fatalf("namespace %v is invalid.", spec.Namespaces[index])
-		} else if spec.Namespaces[index].Type == rspec.UTSNamespace {
+	for index := 0; index < len(spec.Linux.Namespaces); index++ {
+		if !namespaceValid(spec.Linux.Namespaces[index]) {
+			logrus.Fatalf("namespace %v is invalid.", spec.Linux.Namespaces[index])
+		} else if spec.Linux.Namespaces[index].Type == rspec.UTSNamespace {
 			utsExists = true
 		}
 	}
 
-	if !utsExists && hostname != "" {
+	if spec.Platform.OS == "linux" && !utsExists && spec.Hostname != "" {
 		logrus.Fatalf("Hostname requires a new UTS namespace to be specified as well")
```

Comment:
On Wed, May 25, 2016 at 08:10:18PM -0700, 梁辰晔 (Liang Chenye) wrote:
> > -	if !utsExists && hostname != "" {
> > +	if spec.Platform.OS == "linux" && !utsExists && spec.Hostname != "" {
> >  		logrus.Fatalf("Hostname requires a new UTS namespace to be specified as well")
> 
> The PR looks good. But I suggest to change this message `Hostname
> requires a new UTS namespace to be specified as well`.  How about
> `On Linux, hostname requires a UTS namespace to be specified as
> well`?

You want to keep “new” in the error message too, since [1] has:

  Also, when a path is specified, a runtime MUST assume that the setup
  for that particular namespace has already been done and error out if
  the config specifies anything else related to that namespace.

The UTS hostname condition is just a special case of that general rule
[2], and I added it to the spec to avoid surprising people who had
only read the hostname section.

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/config-linux.md#namespaces
[2]: https://github.com/opencontainers/runtime-spec/pull/214